src/libtest/lib.rs

```diff
@@ -872,49 +872,113 @@ pub fn list_tests_console(opts: &TestOpts, tests: Vec<TestDescAndFn>) -> io::Res
     Ok(())
 }
 
-// A simple console test runner
-pub fn run_tests_console(opts: &TestOpts, tests: Vec<TestDescAndFn>) -> io::Result<bool> {
 
-    fn callback<T: Write>(event: &TestEvent, st: &mut ConsoleTestState<T>) -> io::Result<()> {
+/// In multi-threaded modes tests may, of course, start and complete asynchronously.
+///
+/// However our simple line-at-a-time console output presents the illusion they run one
+/// at a time. But, we would still like to show at least one of the tests currently running,
+/// so that slow tests are visible.
+///
+/// This state machine reduces test events to a linear order: at most one can be visibly
+/// pending at a time, and it essentially owns the line of output until it's complete.
+struct SerializedResults<'a, T: Write + 'a> {
+    st: &'a mut ConsoleTestState<T>,
+
+    // The test, if any, that's been printed to the terminal and is waiting for its result
+    // to be available.
+    printed_test: Option<TestName>,
+
+    // Deque of tests that have started, in the order they started.
+    waiting_tests: VecDeque<(TestDesc, NamePadding)>,
+
+    // Map of tests that have completed and not yet been printed.
+    finished_tests: HashMap<TestName, TestResult>,
+}
+
+impl<'a, T: Write> SerializedResults<'a, T> {
+    fn new(st: &'a mut ConsoleTestState<T>) -> SerializedResults<'a, T> {
+        SerializedResults {
+            st: st,
+            printed_test: None,
+            waiting_tests: VecDeque::new(),
+            finished_tests: HashMap::new(),
+        }
+    }
+
+    fn callback(&mut self, event: &TestEvent) -> io::Result<()> {
         match (*event).clone() {
-            TeFiltered(ref filtered_tests) => st.write_run_start(filtered_tests.len()),
-            TeFilteredOut(filtered_out) => Ok(st.filtered_out = filtered_out),
-            TeWait(ref test, padding) => st.write_test_start(test, padding),
-            TeTimeout(ref test) => st.write_timeout(test),
+            TeFiltered(ref filtered_tests) => self.st.write_run_start(filtered_tests.len()),
+            TeFilteredOut(filtered_out) => Ok(self.st.filtered_out = filtered_out),
+            TeWait(ref test, padding) => {
+                self.waiting_tests.push_back((test.clone(), padding));
+                self.advance()
+            },
+            TeTimeout(ref test) => self.st.write_timeout(test),
             TeResult(test, result, stdout) => {
-                st.write_log_result(&test, &result)?;
-                st.write_result(&result)?;
+                self.st.write_log_result(&test, &result)?;
+                self.finished_tests.insert(test.name.clone(), result.clone());
+                self.advance()?;
                 match result {
                     TrOk => {
-                        st.passed += 1;
-                        st.not_failures.push((test, stdout));
+                        self.st.passed += 1;
```

Comment:
This could probably be further cleaned up by separating the printing from the counting in separate observers, but I thought I'd keep this small and get some early feedback.